src/meta/backup_engine.cpp

```diff
@@ -74,9 +71,140 @@ error_code backup_engine::set_block_service(const std::string &provider)
     return ERR_OK;
 }
 
-error_code backup_engine::start() { return ERR_OK; }
+error_code backup_engine::write_backup_file(const std::string &file_name,
+                                            const dsn::blob &write_buffer)
+{
+    dist::block_service::create_file_request create_file_req;
+    create_file_req.ignore_metadata = true;
+    create_file_req.file_name = file_name;
+
+    dsn::error_code err;
+    dist::block_service::block_file_ptr remote_file;
+    _block_service
+        ->create_file(create_file_req,
+                      TASK_CODE_EXEC_INLINED,
+                      [&err, &remote_file](const dist::block_service::create_file_response &resp) {
+                          err = resp.err;
+                          remote_file = resp.file_handle;
+                      })
+        ->wait();
+    if (err != dsn::ERR_OK) {
+        ddebug_f("create file {} failed", file_name);
+        return err;
+    }
+    dassert_f(remote_file != nullptr,
+              "create file {} succeed, but can't get handle",
+              create_file_req.file_name);
+    remote_file
+        ->write(dist::block_service::write_request{write_buffer},
+                TASK_CODE_EXEC_INLINED,
+                [&err](const dist::block_service::write_response &resp) { err = resp.err; })
+        ->wait();
+    return err;
+}
+
+error_code backup_engine::backup_app_meta()
+{
+    dsn::blob app_info_buffer;
+    {
+        zauto_read_lock l;
+        _backup_service->get_state()->lock_read(l);
+        std::shared_ptr<app_state> app = _backup_service->get_state()->get_app(_cur_backup.app_id);
+        if (app == nullptr || app->status != app_status::AS_AVAILABLE) {
+            derror_f("app {} is not available, couldn't do backup now.", _cur_backup.app_id);
+            return ERR_INVALID_STATE;
+        }
+        app_state tmp = *app;
+        // Because we don't restore app envs, so no need to write app envs to backup file.
+        tmp.envs.clear();
```

Comment:
I think it may be better to contain app envs in files, otherwise, app envs will be lost if the old table is dropped. App envs will include many useful data, such as throttle, org, access control etc. If you are sure this information is useless, you can remove it.